src/coreclr/jit/promotiondecomposition.cpp

```diff
@@ -24,13 +25,1181 @@ class DecompositionStatementList
+                addrUse = addr;
+            }
+            else
+            {
+                addrUse = m_compiler->gtCloneExpr(addr);
+                UpdateEarlyRefCount(m_compiler, addrUse);
+            }
+
+            if (offs != 0)
+            {
+                var_types addrType = varTypeIsGC(addrUse) ? TYP_BYREF : TYP_I_IMPL;
+                addrUse            = m_compiler->gtNewOperNode(GT_ADD, addrType, addrUse,
+                                                    m_compiler->gtNewIconNode((ssize_t)offs, TYP_I_IMPL));
+            }
+
+            return addrUse;
+        };
+
+        if (remainderStrategy.Type == RemainderStrategy::FullBlock)
+        {
+            // We will reuse the existing block op's operands. Rebase the
+            // address off of the new local we created.
+            if (m_src->OperIs(GT_BLK))
+            {
+                m_src->AsUnOp()->gtOp1 = grabAddr(0);
+            }
+            else if (m_dst->OperIs(GT_BLK))
+            {
+                m_dst->AsUnOp()->gtOp1 = grabAddr(0);
+            }
+        }
+
+        // If the source involves replacements then do the struct op first --
+        // we would overwrite the destination with stale bits if we did it last.
+        // If the source does not involve replacements then CQ analysis shows
+        // that it's best to do it last.
+        if ((remainderStrategy.Type == RemainderStrategy::FullBlock) && m_srcInvolvesReplacements)
+        {
+            statements->AddStatement(m_compiler->gtNewBlkOpNode(m_dst, m_src));
+
+            if (m_src->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+            {
+                // We will introduce uses of the source below so this struct
+                // copy is no longer the last use if it was before.
+                m_src->gtFlags &= ~GTF_VAR_DEATH;
+            }
+        }
+
+        if (needsNullCheck)
+        {
+            GenTreeIndir* indir = m_compiler->gtNewIndir(TYP_BYTE, grabAddr(0));
+            PropagateIndirFlags(indir, indirFlags);
+            statements->AddStatement(indir);
+        }
+
+        for (int i = 0; i < m_entries.Height(); i++)
+        {
+            const Entry& entry = m_entries.BottomRef(i);
+
+            if (IsHandledByRemainder(entry, remainderStrategy))
+            {
+                assert(entry.FromReplacement != nullptr);
+                JITDUMP("  Skipping dst+%03u <- V%02u (%s); it is up-to-date in its struct local and will be handled "
+                        "as part of the remainder\n",
+                        entry.Offset, entry.FromReplacement->LclNum, entry.FromReplacement->Description);
+                continue;
+            }
+
+            GenTree* dst;
+            if (entry.ToLclNum != BAD_VAR_NUM)
+            {
+                dst = m_compiler->gtNewLclvNode(entry.ToLclNum, entry.Type);
+
+                if (m_compiler->lvaGetDesc(entry.ToLclNum)->lvIsStructField)
+                    UpdateEarlyRefCount(m_compiler, dst);
```

Comment:
This looks a bit odd right now because I'm trying to avoid adding any ref counts to the replacement locals created by this pass. So this logic tracks them correctly only for existing locals and relies on forward sub to give up on the physical promotion locals that have 0 ref count.

I will submit a separate PR after this one to correctly track the ref counts much more uniformly for all the IR created by this pass.